source/fe/mapping_fe.cc

```diff
@@ -909,12 +909,78 @@ MappingFE<dim, spacedim>::transform_real_to_unit_cell(
   const typename Triangulation<dim, spacedim>::cell_iterator &cell,
   const Point<spacedim> &                                     p) const
 {
-  Assert(false, StandardExceptions::ExcNotImplemented());
+  const std::vector<Point<spacedim>> support_points =
+    this->compute_mapping_support_points(cell);
+
+  const double       eps        = 1.e-12 * cell->diameter();
+  const unsigned int loop_limit = 10;
+
+  Point<dim> p_unit;
+
+  unsigned int loop = 0;
+
+  // This loop solves the following problem:
+  // grad_F^T residual + (grad_F^T grad_F + grad_F^T hess_F^T dp) dp = 0
+  // where the term
+  // (grad_F^T hess_F dp) is approximated by (-hess_F * residual)
+  // This is basically a second order approximation of Newton method, where the
+  // Jacobian is corrected with a higher order term coming from the hessian.
+  do
+    {
+      Point<spacedim> mapped_point;
+
+      // Transpose of the gradient map
+      DerivativeForm<1, spacedim, dim> grad_FT;
+      Tensor<1, dim>                   grad_FT_residual;
+      Tensor<2, dim>                   corrected_metric_tensor;
+      DerivativeForm<2, spacedim, dim> hess_FT;
+
+      for (unsigned int i = 0; i < this->fe->n_dofs_per_cell(); ++i)
+        {
+          mapped_point += support_points[i] * this->fe->shape_value(i, p_unit);
+          const auto grad_F_i    = this->fe->shape_grad(i, p_unit);
+          const auto hessian_F_i = this->fe->shape_grad_grad(i, p_unit);
+          for (unsigned int j = 0; j < dim; ++j)
+            {
+              grad_FT[j] += grad_F_i[j] * support_points[i];
+              for (unsigned int l = 0; l < dim; ++l)
+                hess_FT[j][l] += hessian_F_i[j][l] * support_points[i];
+            }
+        }
+
+      // Residual
+      auto residual = p - mapped_point;
+      // Project the residual on the reference coordinate system
+      // to compute the error, and to filter components orthogonal to the
+      // manifold, and compute a 2nd order correction of the metric tensor
+      grad_FT_residual = apply_transformation(grad_FT, residual);
+
+      // Do not invert nor compute the metric if not necessary.
+      if (grad_FT_residual.norm() <= eps)
+        break;
+
+      // Now compute the (corrected) metric tensor
+      for (unsigned int j = 0; j < dim; ++j)
+        for (unsigned int l = 0; l < dim; ++l)
+          corrected_metric_tensor[j][l] =
+            -grad_FT[j] * grad_FT[l] + hess_FT[j][l] * residual;
+
+      // And compute the update
+      auto g_inverse = invert(corrected_metric_tensor);
+      p_unit -= Point<dim>(g_inverse * grad_FT_residual);
+
+      ++loop;
+    }
+  while (loop < loop_limit);
```

Comment:
You dropped the early-exit test for a tolerance. Was this on purpose? This is a relatively expensive function, so it might be useful to check on the size of the residual.

Reply:
it is up there, right after I compute the residual, before I compute the rest. If the residual is lower than tolerance, I just exit with a `break` from the loop.